### DISCUSSION TITLE
HTTP Header Authentication or Single Sign-On (SSO) for netbox with nginx at the front for SSL Termination?

### DISCUSSION THREAD

**Discussion by rakesh-p** [netbox-community/netbox#12349] [Help Wanted!] [answered]
I have setup Netbox using ansible role https://github.com/lae/ansible-role-netbox and would like to configure SSO/Remote authentication. In my setup, i have opensource Nginx at the front (for SSL termination). After some readings, I understood that I cannot use the opensource nginx to act as relay  (OpenID Connect Relying Party) but need something like https://github.com/zmartzone/lua-resty-openidc. This means with simple nginx I cannot use the HTTP Header Authentication method. is this correct? 

So my only option is to use single sign-on authentication via the python-social-auth library. My backend keycloak OIDC/OAUTH2 is in the list of supported backends. My question is will SSO based on python-social-auth work with nginx at the front? I could not find tutorial/guides describing how to configure netbox using python-social-auth. Any inputs/comments is appreciated. thanks

I have seen this https://github.com/netbox-community/netbox/discussions/9635 and looks like people have trouble making SSO to work.

**Comment by candlerb** [comment]
I believe that with Nginx you need Nginx Plus for OIDC authentication, correct.

I use Apache with mod_auth_openidc. This performs the OpenIDC authentication and sets the proxy headers and it works well for me.  For other applications I have used [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy) which also works well; but Apache has the ability to serve the static files which Netbox needs.

For working Apache configs see https://github.com/netbox-community/netbox/discussions/9635#discussioncomment-3057392 - it looks like you found that thread already.

> So my only option is to use single sign-on authentication via the python-social-auth library.

That can be made to work, kind-of, but I couldn't find any way to get it to respect the 'groups' claim. #8579

python-social-auth is an ugly mess :-(

**Comment by rakesh-p** [comment]
Thanks @candlerb for your reply. as nginx opensource doesnt support SSO integration https://www.nginx.com/products/nginx/compare-models/ . I will try apache with mod_auth_openidc and hopefully can get it working.

**Reply by rakesh-p** [reply]
I configured apache with mod_auth_openidc to use netbox.. I am nearly there.. i moved away from ansible installation of netbox to docker version. While the login redirect me to the keycloak and i can authenticate, i get the "431 Request Header Fields Too Large" after authentication. tcpdump between apache and netbox shows all the roles that i configured on keycloak side (OIDC_CLAIM_roles: netbox:superuser,netbox:access,netbox:staff). The only difference with respect to apache configs in https://github.com/netbox-community/netbox/discussions/9635#discussioncomment-3057392 is that i need to add "OIDCCacheShmEntrySizeMax 65536". If i dont add this param, i get the error 
[auth_openidc:error] [pid 235734:tid 139815764133632] [client REDACTED] oidc_cache_shm_set: could not store value since value size is too large (%lu > 45583); consider increasing OIDCCacheShmEntrySizeMax

@candlerb do you remember facing this issue please? thanks

**Reply by candlerb** [reply]
> @candlerb do you remember facing this issue please?

No, but since the error message you get tells you what to do, and making the change fixes the problem, it seems reasonable enough to me. Maybe your OIDC responses include larger volumes of other claims than mine do.

If you still get the 431 error, then use your browser's developer console to inspect the requests going between the browser and Netbox. Possibly you have a large volume of cookies being sent, in which case, clear them in your browser.  You could also try increasing [LimitRequestFieldSize](https://httpd.apache.org/docs/2.4/mod/core.html#limitrequestfieldsize) beyond its default of 8190.

**Reply by candlerb** [reply]
I think Apache gives [400 Bad Request errors](https://www.itsfullofstars.de/2022/02/apache-ah00561-size-of-a-request-header-field-exceeds-server-limit/) for oversized requests, not 431, so this may be a red herring.

I suggest you run a tcpdump to look at the traffic between Apache and gunicorn (netbox):

```
tcpdump -i lo -nn -s0 -A tcp port 8001
```

Then you can see if the request is being passed through by Apache, and it's gunicorn that's giving the 431 response.

If it *is* Apache which is blocking the request with a 431 error, then I found a different proxy instance where I was using these settings:

```
# https://askubuntu.com/questions/340792/size-of-a-request-header-field-exceeds-server-limit-due-to-many-if-none-match-va
# Seems to be needed for grafana
LimitRequestFieldSize 524288

# And this is for the Kafka dashboard which can list a very large number of topics
LimitRequestLine 65536
```

I'm not saying you should need these for Netbox, just that I know I *have* had to tweak these settings for some applications in the past.  Also I suggest you start with **much** lower limits than those, say 32768 and 1024, if you want to experiment.

**Reply by rakesh-p** [reply]
Thanks a lot @candlerb your comment did help me..  as i am a new to keycloak, i troubleshooted the token passed by keycloak and it included attribute called picture which was the culprit. I no longer get 431 error after removing the picture attribute.  Even though i no longer get any error, after the keycloak authentication i get the netbox page but the user is not logged in. It shows login button. I will try to debug myself referring to docs and the response header.

**Reply by rakesh-p** [reply]
After removing the picture attribute from keycloak response, i also dont need to set "OIDCCacheShmEntrySizeMax 65536".

**Reply by candlerb** [reply]
You shouldn't get a login screen if Apache has already authenticated the user. You'll need to check these Netbox settings (remember to restart Netbox if you change them):

```
REMOTE_AUTH_ENABLED = True
REMOTE_AUTH_BACKEND = 'netbox.authentication.RemoteUserBackend'
REMOTE_AUTH_HEADER = 'HTTP_REMOTE_USER'
```

And check with tcpdump whether a "REMOTE_USER" header is being passed (if not, focus your attention there).

**Reply by rakesh-p** [reply]
I made it work now.. i can see after authentication, instead of login button at the top right corner, my username is shown. how does privileges are assigned. I have the following in  netbox config  "REMOTE_AUTH_SUPERUSER_GROUPS = ['netbox:superuser']" and i saw all the 3 roles are passed in the header via tcpdump (OIDC_CLAIM_roles: netbox:superuser,netbox:access,netbox:staff). I understood that if the header contains role 'netbox:superuser', then netbox will provide the superuser privilege. But i can see i am neither admin not superuser. Am i missing some other config to assign privileges  please

**Reply by candlerb** [reply, marked as answer]
Can you show your auth config? The key settings involved are

```
REMOTE_AUTH_GROUP_HEADER = 'HTTP_OIDC_CLAIM_ROLES'
REMOTE_AUTH_GROUP_SYNC_ENABLED = True
REMOTE_AUTH_GROUP_SEPARATOR = ','
```

and it works for me. When you click your username on top-right, does the drop-down menu include an "Admin" option?

Actually, you might also need to create these groups via the Netbox UI:
![image](https://user-images.githubusercontent.com/44789/234862648-8da364c6-116d-4849-a0c1-0e70246249e2.png)

There is a setting `REMOTE_AUTH_AUTO_CREATE_USER = True` to auto-create users, but not one for groups. See #7671

**Reply by rakesh-p** [reply]
My netbox remote auth configuration is same as the one in your reply #9635 
I simply copy pasted it from there.  I do not see Admin in the dropdown though. In other applications like Grafana, I usually see login with username and password or with SSO option. Is it possible to get similar login options with netbox? if yes, then i can login as admin and can create those groups. thanks

**Reply by candlerb** [reply]
You'll have to disable your SSO temporarily (`REMOTE_AUTH_ENABLED = False`) while you sort out the groups.

You can't have a choice of "local login *or* SSO" with this approach - that's the one advantage that the python-social-auth has.

**Reply by rakesh-p** [reply]
yep, it worked now. I disabled remote auth, created groups via netbox UI using local admin account, then enabled remote auth and logged in using OIDC and i am now superuser and admin as expected. thanks a ton for your replies which pointed me in the right direction.
